tests/test_engine.py

```diff
@@ -2243,3 +2244,26 @@ def parent():
         " the same flow."
         in caplog.text
     )
+
+
+@patch(
+    "prefect.utilities.collections.visit_collection",
+    wraps=prefect.utilities.collections.visit_collection,
+)
+@patch("prefect.engine.visit_collection", wraps=prefect.engine.visit_collection)
+async def test_collect_task_run_inputs_respects_quote(
+    mock_outer_visit_collection, mock_recursive_visit_collection
+):
+    # Regression test for https://github.com/PrefectHQ/prefect/pull/10370
+    # This test patches the original `visit_collection` functional call in
+    # `collect_task_run_inputs` and the recursive call inside `visit_collection`
+    # separately. Using `quote` should now keep recursive calls from happening,
+    # as we no longer introspect the input if annotated.
+
+    await collect_task_run_inputs([{"a": 1}, {"b": 2}, {"c": 3}])
+    assert mock_outer_visit_collection.call_count == 1
+    assert mock_recursive_visit_collection.call_count == 9
+
+    await collect_task_run_inputs(quote([{"a": 1}, {"b": 2}, {"c": 3}]))
+    assert mock_outer_visit_collection.call_count == 2
+    assert mock_recursive_visit_collection.call_count == 9
```

Comment:
I had to read this three times before I realized that what you're showing here is that we only get one more outer call and _no more recursive calls_, and I thought the test was wrong.  Can we do this with `reset_mock`s to clarify?

```suggestion

    mock_outer_visit_collection.reset_mock()
    mock_recursive_visit_collection.reset_mock()
    
    await collect_task_run_inputs(quote([{"a": 1}, {"b": 2}, {"c": 3}]))
    assert mock_outer_visit_collection.call_count == 1
    assert mock_recursive_visit_collection.call_count == 0
```

Another way to clarify this could be to make the two cases separate tests with the same setup.

Reply:
you're right, much cleaner, will update!